compiler/runtime/posix.c

```
@@ -64,29 +64,20 @@ void SKIP_posix_close(int64_t fd) {
   }
 }
 
-int64_t SKIP_posix_write(int64_t fd, char *buf) {
-  int rv = write((int)fd, buf, SKIP_String_byteSize(buf));
+int64_t SKIP_posix_write(int64_t fd, char *buf, int64_t len) {
```

Possibly silly question: is there an expectation that this code should work on non-x86/AArch64? For example, the casts in this function will have odd effects on 32 bit platforms (including, I guess, WASM?). But that might not be an issue if they're not expected to be supported.

---

At this point, there really isn't any guarantee anything works beyond amd64/aarch64/wasm32. The runtime could use quite a bit of clean up though, I concede!

---

I think even on wasm32, we have to be careful about the `(int64_t) -> (size_t)` cast(s) as that will do something surprising if someone asks for more than 4GiB, silently truncating. [I've been bitten by this sort of thing in the past so I'm a bit paranoid!]

---

That's a fair point, though `Array`'s size is stored as a 32bit integer, so it shouldn't be an issue (for now).

---

In that case, wouldn't it be better in terms of ease of future debugging to declare the len arg as an int32_t?

---

What I've done in the past is put the casts into a function/macro: it can be infallible for now, and not too verbose, and if this ever becomes an issue in the future it can become a fallible check. [Warning: not only am I paranoid about this but I am also a bore. https://tratt.net/laurie/blog/2021/static_integer_types.html goes into too much detail about these challenges!]

---

> In that case, wouldn't it be better in terms of ease of future debugging to declare the len arg as an int32_t?

My understanding is that while the array length itself is stored in memory as an `int32_t`, it is exposed as an `Int` in (native) skip binaries, which always use `int64_t`.

In any case, the POSIX interface in general, and this file (`posix.c` in particular) are never compiled for wasm.

> What I've done in the past is put the casts into a function/macro

I agree that it would be best (and there is actually a `SkipInt` typedef that is being used sporadically). I'll try to do a clean up pass on the runtime soon™.

---

Ok yeah, makes sense since it's exposed as a skip int.

Sync with me before making a broad pass, I'm accumulating some cleanup diffs for the runtime too.